crates/oxc_linter/src/service/runtime.rs

```diff
@@ -102,105 +191,424 @@ impl Runtime {
+
+        // Sorting paths to make deeper paths appear first.
+        // Consider a typical scenario:
+        //
+        // - src/index.js
+        // - src/a/foo.js
+        // - src/b/bar.js
+        // ..... (thousands of sources)
+        // - src/very/deep/path/baz.js
+        //
+        // All paths above are in `self.paths`. `src/index.js`, the entrypoint of the application, references
+        // almost all the other paths as its direct or indirect dependencies.
+        //
+        // If we construct the module graph starting from `src/index.js`, contents (sources and semantics) of
+        // all these paths must stay in memory (because they are both entries and part of `src/index.js` dependencies)
+        // until the last dependency is processed.
+        // The more efficient way is to start from "leaf" modules: their dependencies are ready earlier, thus we
+        // can run lint on them and then released their content earlier.
+        //
+        // But it's impossible to know which ones are "leaf" modules before parsing even starts. Here we assume
+        // deeper paths are more likely to be leaf modules  (src/very/deep/path/baz.js is likely to have
+        // fewer dependencies than src/index.js).
+        // This heuristic is not always true, but it works well enough for real world codebases.
+        self.paths.par_sort_unstable_by(|a, b| Path::new(b).cmp(Path::new(a)));
+
+        // The general idea is processing entries and their dependencies in groups. We start from a group of entries
+        // that is small enough to hold in memory but big enough to make use of the rayon thread pool. We build the
+        // module graph from this group of entries, run lint on them, drop their content but keep the module
+        // graph, and then move on to the next group.
+        // This size is empirical based on AFFiNE@97cc814a.
+        let entry_group_size = rayon::current_num_threads() * 4;
+
+        // only stores entry modules in current group
+        let mut entry_modules: Vec<EntryModule> = Vec::with_capacity(entry_group_size);
+
+        // downgrade self to immutable reference so it can be shared among spawned tasks.
+        let me: &Self = self;
+
+        // the module graph
+        let mut modules_by_path =
+            FxHashMap::<Arc<OsStr>, SmallVec<[Arc<ModuleRecord>; 1]>>::default();
+
+        // `encountered_paths` prevents duplicated processing.
+        // It is a superset of keys of `modules_by_path` as it also contains paths that are queued to process.
+        let mut encountered_paths =
+            FxHashSet::<Arc<OsStr>>::with_capacity_and_hasher(me.paths.len(), FxBuildHasher);
+
+        let mut module_relationships =
+            Vec::<(Arc<OsStr>, SmallVec<[Vec<(/*specifier*/ CompactStr, Arc<OsStr>)>; 1]>)>::new();
+
+        let (tx_resolve_output, rx_resolve_output) = mpsc::channel::<ModuleProcessOutput>();
```

Comment:
I think it's best to write an algorithm overview here, because there are so many loops and multi-thread logic here.